docs/spec/constructors.rst

```diff
@@ -0,0 +1,492 @@
+        # it should not be evaluated.
+        def __init__(self, x: int):
+            pass
+
+    assert_type(MyClass(), Any)
+
+If the return type of ``__new__`` is not annotated, a type checker may assume
+that the return type is ``Self`` and proceed with the assumption that the
+``__init__`` method will be called.
+
+If the class is generic, it is possible for a ``__new__`` method to override
+the specialized class type and return a class instance that is specialized
+with different type arguments.
+
+  ::
+
+    class MyClass[T]:
+        def __new__(cls, *args, **kwargs) -> "MyClass[list[T]]":
+            ...
+
+    assert_type(MyClass[int](), MyClass[list[int]])
+
+If the ``cls`` parameter within the ``__new__`` method is not annotated, type
+checkers should infer a type of ``type[Self]``. Regardless of whether the
+type of the ``cls`` parameter is explicit or inferred, the type checker should
+bind the class being constructed to the ``cls`` parameter and report any type
+errors that arise during binding.
+
+  ::
+
+    class MyClass[T]:
+        def __new__(cls: "type[MyClass[int]]") -> "MyClass[int]": ...
+
+    MyClass()  # OK
+    MyClass[int]()  # OK
+    MyClass[str]()  # Type Error
+
+
+``__init__`` Method
+===================
+
+After evaluating the ``__new__`` method, a type checker should evaluate the
+``__init__`` method (if applicable) using the supplied arguments. If the class
+is generic and explicitly specialized (or specialized via the ``__new__`` method
+return type), the type checker should partially specialize the ``__init__``
+method using the supplied type arguments. If the class is not explicitly
+specialized, class-scoped type variables should be solved using the supplied
+arguments passed to the constructor call.
+
+This step should be skipped if the class does not define an ``__init__`` method
+and does not inherit an ``__init__`` method from a base class other than
+``object``.
+
+  ::
+
+    class MyClass[T]:
+        def __init__(self, x: T) -> None:
+            ...
+
+    # Constructor calls for specialized classes
+    assert_type(MyClass[int](1), MyClass[int])
+    assert_type(MyClass[float](1), MyClass[float])
+    MyClass[int](1.0)  # Type error
+
+    # Constructor calls for non-specialized classes
+    assert_type(MyClass(1), MyClass[int])
+    assert_type(MyClass(1.0), MyClass[float])
+
+If the ``self`` parameter within the ``__init__`` method is not annotated, type
+checkers should infer a type of ``Self``. Regardless of whether the ``self``
+parameter type is explicit or inferred, a type checker should bind the class
+being constructed to this parameter and report any type errors that arise
+during binding.
+
+  ::
+
+    class MyClass[T]:
+        def __init__(self: "MyClass[int]") -> None: ...
+
+    MyClass()  # OK
+    MyClass[int]()  # OK
+    MyClass[str]()  # Type Error
+
+The return type for ``__init__`` is always ``None``, which means the
```

Comment:
Am I right in thinking that all Python type checkers (sometimes in their more strict modes) will require that the ``__init__`` method return type always be annotated with ``-> None``? That has been my experience and it seems like needless boilerplate.

Can this be made more of an _option_ instead of a requirement specifically for ``__init__``? There may be other magic methods that could also use the same treatment but I have not gone through all of them.